### PR TITLE
Project structure and list products

### PR DESCRIPTION
**Pull request by AndresSan6**

### CODE REVIEW

File: application/src/main/java/com/skytouch/springboot/view/controller/ProductController.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import javax.validation.Valid;
+import java.io.IOException;
+import java.util.List;
+
+@RestController
+public class ProductController {
+
+    private static final Logger log = LoggerFactory.getLogger(ProductController.class);
+    private final Producer producer;
+//    private final Consumer consumer;
```
Comment (Joxebus) [inline application/src/main/java/com/skytouch/springboot/view/controller/ProductController.java]: Remove commented lines.

File: application/src/main/resources/META-INF/resources/WEB-INF/jsp/listProducts.jsp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	<h2>How to fetch data from database in Spring MVC</h2>
+
+
+	<table border="1">
+
+		<th>Name</th>
+		<th>Price</th>
+		<th>Description</th>
+
+		<c:forEach var="listProducts" items="${listProducts}" varStatus="status">
```
Comment (Joxebus) [inline application/src/main/resources/META-INF/resources/WEB-INF/jsp/listProducts.jsp]: The var's name should reflect what it is, in this case you are iterating the `listProducts` so the var should be `product`

File: application/src/main/resources/application.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    username: root
+    password: ${DOCKER_MYSQL_PW}
+    driver-class-name: com.mysql.cj.jdbc.Driver
+  rabbitmq:
+    host: localhost
+    port: 5672
+    username: guest
+    password: guest
+
+jsa:
```
Comment (Joxebus) [inline application/src/main/resources/application.yml]: What means `jsa`?

File: library/src/main/resources/application.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    host: localhost
+    port: 5672
+    username: guest
+    password: guest
+
+jsa:
+  rabbitmq:
+    queue: spring-boot-exchange
+    exchange: spring-boot
+    routingkey: foo.bar.#
```
Comment (Joxebus) [inline library/src/main/resources/application.yml]: Why are you still using `foo.bar.#` routing key, imagine this is an app to be deliver you need to add significative names to your configurations.

File: microservice/pom.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    <parent>
+        <groupId>com.skytouch</groupId>
+        <artifactId>SkyTouch-Task</artifactId>
+        <version>1.0</version>
+    </parent>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
```
Comment (Joxebus) [inline microservice/pom.xml]: What is the propose of this `library` dependency?

File: microservice/src/main/java/main/java/com/skytouch/springboot/view/config/RabbitMqConfig.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        return rabbitTemplate;
+    }
+
+    @Bean
+    public MessageConverter jsonMessageConverter(){
+        return new Jackson2JsonMessageConverter();
+    }
+
+    @Bean
+    public SimpleRabbitListenerContainerFactory jsaFactory(ConnectionFactory connectionFactory,
```
Comment (Joxebus) [inline microservice/src/main/java/main/java/com/skytouch/springboot/view/config/RabbitMqConfig.java]: What is `jsa` meaning I'm confused.

File: microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    private static final Logger log = LoggerFactory.getLogger(ProcedureInvoker.class);
+
+    public ProcedureInvoker(EntityManager entityManager, Producer producer) {
+        this.entityManager = entityManager;
+        this.producer = producer;
+    }
+
+    public void insertProduct(Product product){
+
+        String parameterName = "name";
```
Comment (Joxebus) [inline microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java]: If you are going to use this as constants should be declared at class level as `static final`

File: pom.xml [hunk 8]
```diff
@@ -46,26 +43,11 @@
             <artifactId>spring-boot-starter-data-jpa</artifactId>
         </dependency>
 
-        <dependency>
-            <groupId>mysql</groupId>
-            <artifactId>mysql-connector-java</artifactId>
-            <version>8.0.15</version>
-        </dependency>
-
         <dependency>
```
Comment (Joxebus) [inline pom.xml]: Verify if the this dependencies should go here

File: application/src/main/java/com/skytouch/springboot/view/consumer/Consumer.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    private final ProductController productController;
+
+    public Consumer(ProductController productController) {
+        this.productController = productController;
+    }
+
+    @RabbitListener(queues = "queueListResponse", containerFactory = "jsaFactory")
+    public void receiveListRequest(List resultList) throws IOException {
+        log.info("Product list received");
+//        return resultList;
```
Comment (ragore09) [inline application/src/main/java/com/skytouch/springboot/view/consumer/Consumer.java]: remove commented lines

File: application/src/main/java/com/skytouch/springboot/view/controller/ProductController.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @RequestMapping(value = "/requestListProducts", method = RequestMethod.GET)
+    public void requestListProducts() throws IOException {
+        producer.sendRequestList("request");
+    }
+
+    @RequestMapping(value = "/listProducts", method = RequestMethod.GET)
+    public ModelAndView listProducts(List listProducts) throws IOException {
+//        producer.sendRequestList("request");
```
Comment (ragore09) [inline application/src/main/java/com/skytouch/springboot/view/controller/ProductController.java]: same here about commented lines

File: microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            storedProcedureQuery.execute();
+            log.info("Product inserted: {}", product);
+        } catch (Exception e){ //Cause by Duplicate Entry
+            log.error("Message error: ", e);
+        }
+    }
+
+    public void listProducts(){
+        StoredProcedureQuery storedProcedureQuery = entityManager.createStoredProcedureQuery("list_products");
+//        storedProcedureQuery.registerStoredProcedureParameter("list", Product.class, ParameterMode.OUT);
```
Comment (ragore09) [inline microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java]: remove commented lines

File: application/src/main/java/com/skytouch/springboot/view/config/RabbitMqConfig.java [hunk 12]
```diff
@@ -13,42 +13,69 @@
...
+
+    public String getRoutingkey() {
+        return routingkey;
+    }
+
+    public void setRoutingkey(String routingkey) {
+        this.routingkey = routingkey;
+    }
 
+    //TODO Change this queue to productInsertQueue
```
Comment (ragore09) [inline application/src/main/java/com/skytouch/springboot/view/config/RabbitMqConfig.java]: please avoid commiting commented lines

File: application/src/main/resources/application.yml [hunk 13]
```diff
@@ -6,22 +6,21 @@ spring:
   jpa:
     hibernate:
       ddl-auto: none
-  datasource:
-    url: jdbc:mysql://127.0.0.1:3308/SkyTouchApp
-    username: root
-    password: ${DOCKER_MYSQL_PW}
-    driver-class-name: com.mysql.cj.jdbc.Driver
+#  datasource:
```
Comment (ragore09) [inline application/src/main/resources/application.yml]: why are you commenting this configuration lines?

File: library/src/main/resources/application.yml [hunk 14]
```diff
@@ -13,16 +13,14 @@ spring:
     driver-class-name: com.mysql.cj.jdbc.Driver
   rabbitmq:
     host: localhost
-    port: 5672
+    port: 6789
```
Comment (ragore09) [inline library/src/main/resources/application.yml]: why are you changing the port?

File: microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java [hunk 15]
```diff
@@ -14,6 +14,9 @@
 @Repository
 public class ProcedureInvoker {
 
+    public static final String NAME = "name";
```
Comment (ragore09) [inline microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java]: why are you making all these constants `public` ?  are you planning to use them in other classes?

File: .gitignore [hunk 16]
```diff
@@ -9,4 +9,3 @@ mvnw.cmd
 .mvn/
 jsa-app.log
 *.iml
-*.war
```
Comment (ragore09) [inline .gitignore]: I recommend you checking [GitHub gitignore](https://github.com/github/gitignore) out. Here you'll find really useful templates for multiple programming languages.

File: library/src/main/resources/application.yml [hunk 17]
```diff
@@ -23,4 +23,4 @@ logging:
 rabbitmqconfig:
   queueName: product-insert-request
   exchange: spring-boot
-  routingkey: foo.bar.#
\ No newline at end of file
+  routingkey: foo.bar.#
```
Comment (fjavierponce) [inline library/src/main/resources/application.yml]: use a descriptive name for the queue.

File: application/src/main/resources/application.yml (hunk 3 above)
Comment (AndresSan6) [reply application/src/main/resources/application.yml]: It was for testing, this is already addressed in the next commit

File: library/src/main/resources/application.yml (hunk 4 above)
Comment (AndresSan6) [reply library/src/main/resources/application.yml]: This is addressed in the next commit.

File: microservice/pom.xml (hunk 5 above)
Comment (AndresSan6) [reply microservice/pom.xml]: With this dependency, I am able to use `Product.class` in the Application and Microservice modules.

File: pom.xml (hunk 8 above)
Comment (AndresSan6) [reply pom.xml]: I moved it to the microservice pom.xml

File: library/src/main/resources/application.yml (hunk 14 above)
Comment (AndresSan6) [reply library/src/main/resources/application.yml]: In the expectations for this task (https://docs.google.com/document/d/19_ccXnT_TZlHmy3VSkMl4Ha4PNAzUfxXPZkCwrdvmnQ/edit#) they are asking to listen to message in a non-default port.

File: microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java (hunk 15 above)
Comment (AndresSan6) [reply microservice/src/main/java/main/java/com/skytouch/springboot/view/mysql/ProcedureInvoker.java]: No, thanks for pointing this out. Will change it in the next commit.

File: .gitignore (hunk 16 above)
Comment (AndresSan6) [reply .gitignore]: Thanks, i will check them.

File: library/src/main/java/com/skytouch/springboot/view/Library.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
+package com.skytouch.springboot.view;
```
Comment (fjavierponce) [inline library/src/main/java/com/skytouch/springboot/view/Library.java]: How are you using this library module? if this is only intended to contain common dependencies and classes, I think you shouldn't be creating a spring boot app. A common library should be used as a set of classes common for more than one module, so you can share same resources between your modules.

Comment (AndresSan6) [reply library/src/main/java/com/skytouch/springboot/view/Library.java]: You are right! I fixed this issue in my last commit

**Pull request merged**